### PR TITLE
New merge method in THREE.BufferGeometry.

### PR DESCRIPTION
**Pull request by Wilt**
Merges both indexed and non indexed buffer geometries.

Here a demo: https://jsfiddle.net/wilt/17tzmbqn/

### CODE REVIEW

**Comment by WestLangley** [comment]
If you remove `AmbientLight` from your fiddle, the scene goes black.

Does your merged geometry have normals?

File: src/core/BufferGeometry.js [hunk 1]
```diff
@@ -765,40 +765,75 @@ THREE.BufferGeometry.prototype = {
...
 
-		if ( offset === undefined ) offset = 0;
-
 		var attributes = this.attributes;
 
+		if ( this.index ) {
+
+			var indices = geometry.index.array;
+
+			var offset = attributes[ 'position' ].count;
```
Comment (chrmoritz) [inline src/core/BufferGeometry.js]: This is not always the case. e.g.:
```javascript
/**
 * 2--3
 * | \  |
 * 0--1
 */
var positions = [
    0, 0, 0,
    1, 0, 0,
    0, 1, 0,
    1, 1, 0
  ];

  var indices = [
    0, 1, 2,
    3, 2, 1
  ];
```

File: src/core/BufferGeometry.js [hunk 2]
```diff
@@ -765,40 +765,75 @@ THREE.BufferGeometry.prototype = {
...
+
+		}
+
 		for ( var key in attributes ) {
 
 			if ( geometry.attributes[ key ] === undefined ) continue;
 
-			var attribute1 = attributes[ key ];
-			var attributeArray1 = attribute1.array;
+			attributes[ key ].array = Float32ArrayConcat( attributes[ key ].array, geometry.attributes[ key ].array );
```
Comment (chrmoritz) [inline src/core/BufferGeometry.js]: Also (custom) `BufferAttribute` are not limited to `Float32Array`, see: https://github.com/mrdoob/three.js/blob/master/src/core/BufferAttribute.js#L307

**Comment by Wilt** [comment]
@WestLangley No it doesn't because I didn't define any :D 

But the merge works for them. I updated **[my fiddle](https://jsfiddle.net/wilt/17tzmbqn/4/)** and added normals now, to show that it works. Ambient light is off now...

**Comment by Wilt** [comment]
@chrmoritz You are very right in your comment. I guess I should think about how to add support for all those different ArrayTypes. But I rather work on that if I know there is some interest for this PR at all...

**Comment by chrmoritz** [comment]
Maybe move your helper methods to `THREE.BufferAttribute` and implement is like this:
```javascript
THREE.BufferAttribute.prototype.merge = function(attribute){
    var attributeArray1 = this.array;
    var attributeArray2 = attribute.array;
    if ( attributeArray1 .constructor !== attributeArray2 .constructor || this.itemSize !== attribute.itemSize ){
        console.error( 'THREE.BufferAttribute.merge(): type mismatch', attribute );
    }
    var firstLength = attributeArray1.length;
    var result = new attributeArray1.array.constructor( firstLength + attributeArray2.length );
    result.set( attributeArray1 );
    result.set( attributeArray2, firstLength );
    this.array = result;
    return this;
}
```

**Comment by Wilt** [comment]
I implemented a new merge method inside `THREE.BufferAttribute`, very much like @chrmoritz suggested. You can see it [here](https://github.com/Wilt/three.js/blob/dev/src/core/BufferGeometry.js#L768-L828).

Merge will throw errors in all non matching merge operations. Meaning 

1) When merging non-indexed and indexed
2) When merging geometries with different (non corresponding attribute members.

**[Here is an updated fiddle](https://jsfiddle.net/wilt/17tzmbqn/6/)** where these new methods are used.

**Comment by Wilt** [comment]
I updated the docs too. I wanted to refer to the `.toNonIndexed` method, but this one wasn't in the docs, so I added it. I hope my docs update is fine, since I am not able to see the updated documentation live in the browser window (if this is possible, please tell me how to do that).

**Comment by chrmoritz** [comment]
Look good, except that your current implementation still doesn't support BufferGeometries with custom BufferAttributes because of the hard coded `attributes = [ 'position', 'normal', 'color', 'uv' ]`.
That would be a breaking change because the current `THREE.BufferGeometry.prototype.merge()` implementations does support merging (non-indexed) BufferGeometries with custom BufferAttributes.

**Comment by Wilt** [comment]
@chrmoritz  Good point, I didn't know custom attributes were possible, but [this custom attribute example](http://threejs.org/examples/webgl_buffergeometry_custom_attributes_particles.html) shows it this.
I updated once more. Should work for custom now too...

**Comment by WestLangley** [comment]
@Wilt Your test example is very simple. Can you provide test with more complicated geometries?

One that comes to mind is two `SphereBufferGeometry`s each having `index.array` as a `Uint16Array` but when merged they require a `Uint32Array`.

Also what happens if the attributes match but are in a different order?

In any case, failure must be graceful -- perhaps just returning the geometry unmodified with a console warning.

**Comment by drcmda** [comment]
I seem to not be able to merge larger geometries. WebGL throws buffer-overflow errors. It works for smaller geometries. I found conflicting information regarding this - at one place saying webGL is limited to UInt6 and 65535 vertices, but then there's OES_element_index_uint which should in theory allow UInt32 and bump vertices up to 4294967296. My browser supports that for sure. In any case, documents also mix indicies and vertices so i'm not sure what exactly is limited.

Is there a definitive answer for this?  Has anyone tried merging large BufferGeometries with this?

File: src/core/BufferAttribute.js [hunk 3]
```diff
@@ -298,6 +298,26 @@ THREE.BufferAttribute.prototype = {
...
+
+		var result, first = this.array, second = attribute.array, length = first.length;
+
+		if ( first.constructor !== second.constructor || this.itemSize !== attribute.itemSize ) {
+
+			console.error( 'THREE.BufferAttribute.merge(): type mismatch', attribute );
+
+		}
+
+		result = new first.constructor( length + second.length );
```
Comment (SereznoKot) [inline src/core/BufferAttribute.js]: What if two attributes together will overflow Uint16Array?
Maybe you should do something like this:
`result = new ( ( length + second.length ) > 65535 ? Uint32Array : Uint16Array )( length + second.length );`

**Comment by Wilt** [comment]
Sorry to bother you with this @mrdoob but I rebased my dev fork and now I noticed that [in r70 you already added `THREE.BufferGeometry` `merge` method](https://github.com/mrdoob/three.js/releases/tag/r70)

Meaning this PR can be closed right...

**Comment by mrdoob** [comment]
Yeah, I'm not sure this is good practice... Years ago it was fine with `Geometry` because an `Array` can keep growing, but `TypedArray`s can't...

**Comment by Wilt** [comment]
OK, I understand your concerns about such functionality, but [there seems to be a demand](http://stackoverflow.com/search?q=%5Bthree.js%5D+merge+buffergeometry) and it seems people are confused about the working of the current merge method. 

If you change your mind and I can help in any way just let me know.

**Comment by h1Modeling** [comment]
Hi, 
I don´t know if this is interesting to you but I like to add a personal comment on this topic:

I´ve developed a building configurator (www.hallenkonfigurator.com). I´m using a lot of parametric Geometries (currently THREE.Geometry). These building parts (doors windows etc.) use a lot of the Geometry.merge functionality. 
I have decided in the past to move to BufferGeometry to gain speed improvements at geometry creation. As Wilt pointed out I was confused, that BufferGeometry.merge is something different than Geometry.merge. ( Maybe there should be different names to visualize that there is different functionality behind?)
To solve my problem I´ve introduced a new class (BufferData) which holds vertex, uvs and group data in standard arrays. This class also has some basic modification functionality (translate, move, mirror, merge). After I have finished all my vertex modification and merges I then create a BufferGeometry from it and add it to the scene.
This also means that I don´t have to take care of disposing my geometry parts, because it´s just pure data which THREE don´t know about.

P.S. a big thanks to @mrdoob and the other contributors - your tool helped me really a lot!

**Comment by kylegmaxwell** [comment]
I also ran into this issue. I wanted a merge function that behaves like geometry.merge. I recently wrote a function that merges a list of buffer geometry by merging their attributes and creating a new geometry. I am happy to contribute it if you are interested.

**Comment by WestLangley** [comment]
@kylegmaxwell I think your utility would be worth considering if you would be willing to file a PR.

**Comment by kylegmaxwell** [comment]
@WestLangley Thanks, I will put it up once I get approval from my employer.

File: src/core/BufferAttribute.js (hunk 3 above)
Comment (donmccurdy) [reply src/core/BufferAttribute.js]: Uint16Array values are capped at 65535, not their length, so this should be fine.

**Pull request closed without merging**